### ISSUE TITLE
No local inference for array component types

### ISSUE THREAD

**Issue by kelloggm** [typetools/checker-framework#3179]
Array component types on the left-hand side of assignments are not inferred, even though inference succeeds for the right-hand side. This happens for all type systems; my examples use the Index Checker (to distinguish this from #599). Consider the following program:

```
// Test local inference of array constant types

import org.checkerframework.checker.index.qual.*;

class ArrayInfer {
    void testIndex() {
	// I expect nn_annotated and nn_unannotated to have the same
	// (inferred) type, because they're both effectively final
	// and both have the same initializer.
	@NonNegative int[] nn_annotated = new int[] { 0, 5, 3 };
	int[] nn_unannotated = new int[] { 0, 5, 3 };

	// OK
	@NonNegative int[] x = nn_annotated;

	// unexpected error here
	@NonNegative int[] y = nn_unannotated;
    }
}
```

Both arrays are effectively final and have the same initializer. I would therefore expect local type inference to infer that the component type of `nn_unannotated` is `@NonNegative`. However, running `javacheck -processor index ArrayInfer.java` yields:

```
ArrayInfer.java:17: error: [assignment.type.incompatible] incompatible types in assignment.
	@NonNegative int[] y = nn_unannotated;
	                       ^
  found   : @LowerBoundUnknown int @LowerBoundUnknown []
  required: @NonNegative int @LowerBoundUnknown []
1 error
```

By contrast, if I change the code so that this is 3 (unannotated) integers instead of a 3-element integer array, the types are inferred as expected.

I think the problem here is that local type inference doesn't attempt to update the component type, even if the variable is effectively final and there is an initializer.

**Comment by smillst** [comment]
This is the expected behavior.
1. `@NonNegative int[]` isn't a subtype of `@Unknown int[]`, so refinement isn't possible.
2. Dataflow only stores the primary annotation of each annotated type.

If you want to implement this for effectually final local variables, you should be able to do so in a `TreeAnnotator` in the `visitVariable` method.

**Comment by kelloggm** [comment]
> If you want to implement this for effectually final local variables, you should be able to do so in a TreeAnnotator in the visitVariable method.

That was also my first thought. I can modify the type in `visitVariable`, but dataflow doesn't propagate the component type. So, in the example above, I can store `@NonNegative int @Unknown []` in the `AnnotatedTypeMirror` argument to `visitVariable` when visiting `int[] nn_unannotated = new int[] { 0, 5, 3 };`, but then at the assignment later (`@NonNegative int[] y = nn_unannotated;`), the type of the rhs is still `@Unknown int @Unknown []`.

My question is about how (if it is possible) to modify dataflow so that modifying the component type in `visitVariable` will cause the rhs of the assignment to have the expected type.

**Comment by smillst** [comment]
>My question is about how (if it is possible) to modify dataflow so that modifying the component type in visitVariable will cause the rhs of the assignment to have the expected type.

To implement this using dataflow, you will have to change `CFValue` class to store the annotation on the component type.  This would also involve changing `CFAbstractValue#leastUpperBound` and `CFAbstractValue#mostSpecific` to take into account the annotations on the component type.

KeyFor Checker copies type argument annotations from the RHS to the LHS.  See `org.checkerframework.checker.nullness.KeyForPropagationTreeAnnotator#visitVariable`.  You should be able to do something similar for arrays.
